### PR TITLE
Move remaining "Encode*" methods into Writer

### PR DESCRIPTION
**Pull request by rsgowman**
These are the firestore specific encoders.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 1]
```diff
@@ -36,16 +36,22 @@ namespace {
...
-
 std::map<std::string, FieldValue> DecodeObject(pb_istream_t* stream);
 
 /**
  * Docs TODO(rsgowman). But currently, this just wraps the underlying nanopb
  * pb_ostream_t.
+ *
+ * Possible future change: Currently, this class has generic methods that would
+ * apply to any nanopb based protos, but also methods that only apply to
+ * firestore protos. It may be desireable to split these apart to keep these
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Small nit: s/desireable/desirable

**Comment by rsgowman** [comment]
Yeah, I wasn't sure about this one either. I was originally thinking some sort of split such that Writer remains generic, and then extend/encapsulate it with a FirestoreWriter which has knowledge of the firestore specific messages. But upon further reflection, this is essentially what Serializer is. I'll abandon this PR.

> EncodeFieldsEntry does belong on Writer, but seems like it's written at the wrong level of abstraction. It should take as arguments the string key and a function to encode the nested message.

I think I disagree with this. The FieldsEntry is (wave hands) effectively another firestore specific message, not really much different from 'Object' or 'FieldValue'. It's certainly firestore specific. (i.e. it won't work for a generic pair<string, T> since it needs to know the tags.)

We could do a 
```
EncodeMapEntryMessage(keytag, valuetag, keywiretype, valuewiretype, function encodekey, function encodevalue)
```
or possibly
```
EncodeMapStringToTEntryMessage(keytag, string key, valuetag, valuewiretype, function encodevalue)
```

I'll play around with that a bit and see if it yields anything useful.

**Comment by wilhuff** [comment]
Re: EncodeFieldsEntry I see what you mean: I had missed that the tags were specific to that map type, so agreed that doesn't belong on Writer either.

The kind of functions you propose as alternatives don't get me super excited either so let's just leave EncodeFieldsEntry as is (but on the Serializer)

**Comment by rsgowman** [comment]
> The kind of functions you propose as alternatives don't get me super excited either so let's just leave EncodeFieldsEntry as is (but on the Serializer)

Yup. I tried it out. The cure was far worse than the disease.

**Pull request closed without merging**